In [2]:
import numpy as np
import pandas as pd

### Question 1
What's the version of NumPy that you installed?

In [3]:
np.__version__

'1.21.5'

In [4]:
df = pd.read_csv('data.csv')

In [5]:
df

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

### Question 2
How many records are in the dataset?

- 11914 rows

### Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

- Chevrolet, Ford, Volkswagon

In [6]:
df.Make.value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

### Question 4

What's the number of unique Audi car models in the dataset?

- 34

In [7]:
df_audi = df[df.Make == 'Audi']
df_audi.Model.nunique()

34

### Question 5
How many columns in the dataset have missing values?

- 5


In [8]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

### Question 6
- Find the median value of "Engine Cylinders" column in the dataset. <br>
- Next, calculate the most frequent value of the same "Engine Cylinders". <br>
- Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step. <br>
- Now, calculate the median value of "Engine Cylinders" once again.

In [9]:
#median
df['Engine Cylinders'].median()

6.0

In [10]:
#most frequent value
df['Engine Cylinders'].value_counts()

4.0     4752
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: Engine Cylinders, dtype: int64

In [21]:
#most frequent value
engine_cylinders_mode = df['Engine Cylinders'].mode()
engine_cylinders_mode 

0    32.0
Name: Engine Cylinders, dtype: float64

In [13]:
df['Engine Cylinders'][df['Engine Cylinders'].isnull()]

1983   NaN
1984   NaN
3716   NaN
3717   NaN
3718   NaN
3719   NaN
5778   NaN
5779   NaN
5780   NaN
8373   NaN
8695   NaN
8696   NaN
8697   NaN
8698   NaN
8699   NaN
8700   NaN
8701   NaN
8702   NaN
8703   NaN
8704   NaN
8705   NaN
8706   NaN
8707   NaN
8708   NaN
8709   NaN
8710   NaN
8711   NaN
8712   NaN
8713   NaN
8714   NaN
Name: Engine Cylinders, dtype: float64

In [34]:
df['Engine Cylinders'] = df['Engine Cylinders'].fillna(4)

In [35]:
df['Engine Cylinders'][df['Engine Cylinders'].isnull()]

Series([], Name: Engine Cylinders, dtype: float64)

In [40]:
df['Engine Cylinders'].value_counts()

4.0     4782
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: Engine Cylinders, dtype: int64

In [41]:
df['Engine Cylinders'].median()

6.0

The median is still 6

### Question 7
- Select all the "Lotus" cars from the dataset.
- Select only columns "Engine HP", "Engine Cylinders".
- Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
- Get the underlying NumPy array. Let's call it X.
- Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
- Invert XTX.
- Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
- Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
- What's the value of the first element of w?

In [42]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [51]:
df_lotus = df[df.Make == 'Lotus'][[ "Engine HP", "Engine Cylinders"]]
df_lotus

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0
3917      189.0               4.0
3918      217.0               4.0
3919      217.0               4.0
4216      350.0               8.0
4217      350.0               8.0
4218      350.0               8.0
4257      400.0               6.0
4258      400.0               6.0
4259      276.0               6.0
4260      276.0               6.0
4261      276.0               6.0
4262      345.0               6.0
4263      345.0               6.0
4264      276.0               6.0
4265      345.0               6.0
4266      276.0               6.0
4267      345.0               6.0
4268      276.0               6.0
4292      257.0               4.0
4293      240.0               4.0
4294      240.0               4.0
4295      240.0               4.0
4296      257.0               4.0
4297      257.0               4.0

In [54]:
df_lotus = df_lotus.drop_duplicates()
df_lotus

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [57]:
X = np.array(df_lotus)
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [58]:
XTX = X.T.dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [59]:
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

In [60]:
XTX_inv = np.linalg.inv(XTX)

In [66]:
w = XTX_inv.dot(X.T).dot(y)

In [67]:
w[0]

4.594944810094579

The answer is 4.594944810094579